crates/stackable-operator/src/validation.rs

```diff
@@ -26,121 +27,202 @@ const RFC_1123_SUBDOMAIN_MAX_LENGTH: usize = 253;
 const RFC_1123_LABEL_MAX_LENGTH: usize = 63;
 
 const RFC_1035_LABEL_FMT: &str = "[a-z]([-a-z0-9]*[a-z0-9])?";
-const RFC_1035_LABEL_ERR_MSG: &str = "a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character";
+const RFC_1035_LABEL_ERROR_MSG: &str = "a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character";
 
 // This is a label's max length in DNS (RFC 1035)
 const RFC_1035_LABEL_MAX_LENGTH: usize = 63;
 
+// Technically Kerberos allows more realm names
+// (https://web.mit.edu/kerberos/krb5-1.21/doc/admin/realm_config.html#realm-name),
+// however, these are embedded in a lot of configuration files and other strings,
+// and will not always be quoted properly.
+//
+// Hence, restrict them to a reasonable subset. The convention is to use upper-case
+// DNS hostnames, so allow all characters used there.
+const KERBEROS_REALM_NAME_FMT: &str = "[-.a-zA-Z0-9]+";
+const KERBEROS_REALM_NAME_ERROR_MSG: &str =
+    "Kerberos realm name must only contain alphanumeric characters, '-', and '.'";
+
 // Lazily initialized regular expressions
-static RFC_1123_SUBDOMAIN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+pub(crate) static RFC_1123_SUBDOMAIN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
     Regex::new(&format!("^{RFC_1123_SUBDOMAIN_FMT}$"))
         .expect("failed to compile RFC 1123 subdomain regex")
 });
 
+static RFC_1123_LABEL_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(&format!("^{RFC_1123_LABEL_FMT}$")).expect("failed to compile RFC 1123 label regex")
+});
+
 static RFC_1035_LABEL_REGEX: LazyLock<Regex> = LazyLock::new(|| {
     Regex::new(&format!("^{RFC_1035_LABEL_FMT}$")).expect("failed to compile RFC 1035 label regex")
 });
 
-/// Returns a formatted error message for maximum length violations.
-fn max_len_error(length: usize) -> String {
-    format!("must be no more than {length} characters")
-}
+pub(crate) static KERBEROS_REALM_NAME_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(&format!("^{KERBEROS_REALM_NAME_FMT}$"))
+        .expect("failed to compile Kerberos realm name regex")
+});
 
-/// Returns a formatted error message for regex violations.
-///
-/// # Arguments
-///
-/// * `msg` - this is the main error message to return
-/// * `fmt` - this is the regular expression that did not match the input
-/// * `examples` - are optional well, formed examples that would match the regex
-fn regex_error(msg: &str, fmt: &str, examples: &[&str]) -> String {
-    if examples.is_empty() {
-        return format!("{msg} (regex used for validation is '{fmt}')");
-    }
+type Result<T = (), E = Errors> = std::result::Result<T, E>;
```

Comment:
note: This might be a little confusion, because most local `Result` type aliases use `type Result<T = (), E = Error> = std::result::Result<T, E>`. We might want to indicate that this type wraps a list of errors.

Reply:
That's fair, on the other `Errors` is "the" default error type that we expose outwards from the module.

Reply:
Oh yeah right, then let's leave it as is.